File: 3R/2015_09_LongestIncreasingSubsequence.cpp
#include <iostream>

using namespace std;

int LIS(int n, int arr[]){
    
    int length[n]; //記錄每一個數字的LIS長度
    for(int i=0; i<n; i++){
        length[i] = 1; //預設每一個的長度都是 1，也就是只有他自己一個
    }

    int L = 0;
    for(int i=0; i<n; i++){
        for(int j=0; j<i; j++){
            if(arr[j]<arr[i]) length[i] = max(length[i],length[j]+1);
        }
        if(L < length[i]) L = length[i];
    }
    
    return L;
}

int main(){
    int n, ans=0;
    while(cin >> n && n){
        int arr[n];
        for(int i=0; i<n; i++){
            cin >> arr[i];
        }
        ans = LIS(n, arr);
        cout << ans << endl;
    }
}

File: 3R/2015_09_LongestIncreasingSubsequence.java
import java.util.*;

class Main{
	public static void main(String[] args){
		Scanner sc = new Scanner(System.in);
        while(sc.hasNext()){
            int n = sc.nextInt();
            int arr[] = new int[n];
            for(int i=0; i<n; i++){
                arr[i] = sc.nextInt();
            }
            int ans = LIS(n, arr);
            System.out.println(ans);
        }
	}

    public static int LIS(int n, int arr[]){
        int length[] = new int[n];
        if(n==0) return 0;
        Arrays.fill(length, 1); //把整個陣列填滿 1

        int longest = 0;

        for(int i=0; i<n; i++){
            for(int j=i+1; j<n; j++){
                if(arr[i]<arr[j])
                    length[j] = Math.max(length[j], length[i]+1);
            }
            if(longest < length[i]) longest = length[i];
        }
        return longest;
    }
}